src/main/scala/org/constellation/snapshotstreaming/SnapshotProcessor.scala

```diff
@@ -0,0 +1,141 @@
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait SnapshotProcessor[F[_]] {
+  val runtime: Stream[F, Unit]
+}
+
+object SnapshotProcessor {
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider: Random](configuration: Configuration): Resource[F, SnapshotProcessor[F]] =
+    for {
+      client <- EmberClientBuilder
+        .default[F]
+        .withTimeout(configuration.httpClientTimeout)
+        .withIdleTimeInPool(configuration.httpClientIdleTime)
+        .build
+      opensearchDAO <- OpensearchDAO.make[F](configuration.opensearchUrl)
+      s3DAO <- S3DAO.make[F](configuration)
+      l0GlobalSnapshotClient = L0GlobalSnapshotClient.make[F](client)
+      l0ClusterStorage <- Resource.eval {
+        Ref.of[F, NonEmptyMap[PeerId, L0Peer]](configuration.l0Peers).map(L0ClusterStorage.make(_))
+      }
+      lastSnapshotStorage <- Resource.eval {
+        FileBasedLastGlobalSnapshotStorage.make[F](configuration.lastSnapshotPath)
+      }
+      l0Service = L0Service.make[F](l0GlobalSnapshotClient, l0ClusterStorage, lastSnapshotStorage, configuration.pullLimit.some)
+      requestBuilder = UpdateRequestBuilder.make[F](configuration)
+    } yield make(configuration, lastSnapshotStorage, l0Service, opensearchDAO, s3DAO, requestBuilder)
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider: Random](
+    configuration: Configuration,
+    lastGlobalSnapshotStorage: LastGlobalSnapshotStorage[F],
+    l0Service: L0Service[F],
+    opensearchDAO: OpensearchDAO[F],
+    s3DAO: S3DAO[F],
+    updateRequestBuilder: UpdateRequestBuilder[F]
+  ): SnapshotProcessor[F] = new SnapshotProcessor[F] {
+
+    val logger = Slf4jLogger.getLogger[F]
+
+    private def prepareAndExecuteBulkUpdate(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
+      Clock[F].realTime
+        .map(d => new Date(d.toMillis))
+        .flatMap(updateRequestBuilder.bulkUpdateRequests(snapshot, _))
+        .flatMap(_.traverse(br => opensearchDAO.sendToOpensearch(bulk(br).refreshImmediately)))
+        .flatMap(_ => logger.info(s"Snapshot ${snapshot.ordinal.value.value} (hash: ${snapshot.hash.show.take(8)}) sent to opensearch."))
+        .void
+
+    private def process(snapshot: Hashed[GlobalSnapshot]): F[Unit] =
+      lastGlobalSnapshotStorage.get.flatMap {
+        case Some(last) if Validator.isNextSnapshot(last, snapshot) =>
+          s3DAO.uploadSnapshot(snapshot) >>
+            prepareAndExecuteBulkUpdate(snapshot) >>
+            lastGlobalSnapshotStorage.set(snapshot)
+
+        case Some(last) =>
+          logger.warn(s"Pulled snapshot doesn't form a correct chain, ignoring! Last: ${GlobalSnapshotReference
+              .fromHashedGlobalSnapshot(last)} pulled: ${GlobalSnapshotReference.fromHashedGlobalSnapshot(snapshot)}")
+
+        case None if configuration.initialSnapshot.exists(initial => initial.hash === snapshot.hash && initial.ordinal === snapshot.ordinal) =>
+          s3DAO.uploadSnapshot(snapshot) >>
```

Comment:
Should we handle the situation when one of these fail and just retry again on another pull? The question is how will s3 or opensearch behave when the file already exists or partial update to indexes has been already made? @tbekas @kpudlik @kriscieslak @marcinwadon

Reply:
Going to handle the failure and retry again.

Reply:
One of these you mean s3 succeed but opensearch not? imo we should retry. we should definitely not process next ordinals

Reply:
Yeah I meant if one of these fails, should we pull the same snapshot again and try to upload to s3/opensearch that might have the snapshot uploaded or partially uploaded already. Anyway I tested it locally with dockerized opensearch and localstack (mimicks aws) and resending doesn't throw any errors so we can retry. One thing that could happen with retry like this is that if at this particular ordinal the cluster forked and created two snapshots we may on the subsequent tries download any of these snapshots and hence upload parts/everything of both to s3/opensearch. Though this scenario seems to be of little probability. Before handling the error here the whole snapshot-streaming app would stop.